64_MergeSort/main.cpp

```diff
@@ -0,0 +1,368 @@
+
+				queue->startCapture();
+				queue->submit(submitInfos);
+				queue->endCapture();
+
+			}
+			const ISemaphore::SWaitInfo wait_infos[] = { {
+		.semaphore = m_phaseSemaphore.get(),
+		.value = monotonicallyIncreasingCounter,
+	}, };
+
+			m_device->blockForSemaphores(wait_infos);
+		}
+
+		// Now, get the current pointer to output buffer (the output may be in either input or output buffer).
+		// Perform merge sort on the CPU to test whether GPU compute version computed the result correctly.
+		auto outputBufferData = reinterpret_cast<const int32_t*>(m_outputBufferAllocation.memory->getMappedPointer());
+
+		if (numberOfPhases % 2 == 0)
+		{
+			outputBufferData = reinterpret_cast<const int32_t*>(m_inputBufferAllocation.memory->getMappedPointer());
+		}
+
+		// Sorted input (which should match the output buffer).
+		std::sort(inputBufferData.begin(), inputBufferData.end());
+
+		for (auto i = 0; i < NumberOfElementsToSort; i++)
+		{
+			printf("output buffer -> %d input buffer -> %d\n", outputBufferData[i], inputBufferData[i]);
+			if (outputBufferData[i] != inputBufferData[i])
+			{
+				return logFail("%d != %d\n", outputBufferData[i], inputBufferData[i]);
+			}
+		}
+
+		printf("Merge sort succesfull!\n");
+
+		m_device->waitIdle();
+		return true;
+	}
+
+	// Platforms like WASM expect the main entry point to periodically return control, hence if you want a crossplatform app, you have to let the framework deal with your "game loop"
+	void workLoopBody() override {}
+
+	// Whether to keep invoking the above. In this example because its headless GPU compute, we do all the work in the app initialization.
+	bool keepRunning() override { return false; }
+
+private:
+	uint32_t m_computeQueueFamily{};
+
+	// Note : In this implementation of merge sort, the buffers are 'swapped' at the end of each phase (i.e input / output buffer of previous phase becomes the output / input of current phase).
+	// The names reflect the fact that initially the input data is present in m_inputBuffer.
+	smart_refctd_ptr<IGPUBuffer> m_outputBuffer{};
+	smart_refctd_ptr<IGPUBuffer> m_inputBuffer{};
+
+	// Allocator is a interface to anything that can dish out free memory range to bind to a buffer or image.
+	nbl::video::IDeviceMemoryAllocator::SAllocation m_outputBufferAllocation{};
+	nbl::video::IDeviceMemoryAllocator::SAllocation m_inputBufferAllocation{};
+
+	// Semaphore is used because before completion of merge sort phase X, phase X + 1 cannot start.
+	smart_refctd_ptr<ISemaphore> m_phaseSemaphore{};
+
+	smart_refctd_ptr<IGPUDescriptorSetLayout> m_descriptorSetLayout{};
+
+	// There are 2 descriptor sets for merge sort. One has input buffer / output buffer at binding 0, 1, and one has the inverse.
+	// This is because in each merge sort phase, the inputs and output buffers are *swapped*.
+	smart_refctd_ptr<nbl::video::IGPUDescriptorSet> m_descriptorSet1{};
+	smart_refctd_ptr<nbl::video::IGPUDescriptorSet> m_descriptorSet2{};
```

Comment:
you don't need any descriptor sets, just use BDA for the inputs

Reply:
Got it, will use BDA over descriptor sets.